File: Dev/unitTest_Engine_cpp_gtest/FamilySystem/ChildManagement.cpp
#include <gtest/gtest.h>
#include <Altseed.h>
#include "../EngineTest.h"

using namespace std;
using namespace asd;

class FamilySystem_ChildManagement : public EngineTest
{
	class SpinObject : public TextureObject2D
	{
	public:
		typedef shared_ptr<SpinObject> Ptr;

	protected:
		void OnUpdate()
		{
			SetAngle(GetAngle() + 2);
		}
	};

	class Family
	{
		SpinObject::Ptr parent;
		SpinObject::Ptr child;
		SpinObject::Ptr childNotAdded;
		Vector2DF position;
		ChildManagementMode::Flags managementMode;

	public:
		Family(Vector2DF position, ChildManagementMode::Flags managementMode)
			: managementMode(managementMode)
			, position(position)
		{
		}

		void Initialize()
		{
			parent = make_shared<SpinObject>();
			child = make_shared<SpinObject>();
			childNotAdded = make_shared<SpinObject>();

			auto texture = Engine::GetGraphics()->CreateTexture2D(ToAString("Data/Texture/Cloud1.png").c_str());
			parent->SetTexture(texture);
			parent->SetCenterPosition(texture->GetSize().To2DF() / 2);
			parent->SetScale(Vector2DF(0.5f, 0.5f));
			parent->SetPosition(position);

			child->SetTexture(texture);
			child->SetCenterPosition(texture->GetSize().To2DF() / 2);
			child->SetScale(Vector2DF(0.5f, 0.5f));
			child->SetPosition(position + Vector2DF(0, 100));

			childNotAdded->SetTexture(texture);
			childNotAdded->SetCenterPosition(texture->GetSize().To2DF() / 2);
			childNotAdded->SetScale(Vector2DF(0.5f, 0.5f));
			childNotAdded->SetPosition(position + Vector2DF(0, 200));

			parent->AddChild(child, managementMode, ChildTransformingMode::Nothing);
			parent->AddChild(childNotAdded, managementMode, ChildTransformingMode::Nothing);

			Engine::AddObject2D(parent);
			if ((managementMode & ChildManagementMode::RegistrationToLayer) == 0)
			{
				Engine::AddObject2D(child);
			}
		}

		SpinObject::Ptr& GetParent()
		{
			return parent;
		}
	};

public:
	FamilySystem_ChildManagement(bool isOpenGLMode)
		: EngineTest(asd::ToAString("ChildManagement"), isOpenGLMode, 120)
		, count(0)
	{
	}

protected:
	shared_ptr<Family> nothing, isDrawnChanged, isUpdatedChanged, registrationToLayerChanged, vanished;
	int count;

	void OnStart()
	{
		nothing = make_shared<Family>(Vector2DF(70, 70), ChildManagementMode::Nothing);
		isDrawnChanged = make_shared<Family>(Vector2DF(190, 70), ChildManagementMode::IsDrawn);
		isUpdatedChanged = make_shared<Family>(Vector2DF(310, 70), ChildManagementMode::IsUpdated);
		registrationToLayerChanged = make_shared<Family>(Vector2DF(430, 70), ChildManagementMode::RegistrationToLayer);
		vanished = make_shared<Family>(Vector2DF(550, 70), ChildManagementMode::Vanishment);

		nothing->Initialize();
		isDrawnChanged->Initialize();
		isUpdatedChanged->Initialize();
		registrationToLayerChanged->Initialize();
		vanished->Initialize();
	}

	void OnUpdating()
	{
		++count;
		if (count == 60)
		{
			isDrawnChanged->GetParent()->SetIsDrawn(false);
			isUpdatedChanged->GetParent()->SetIsUpdated(false);
			Engine::RemoveObject2D(registrationToLayerChanged->GetParent());
			vanished->GetParent()->Dispose();
		}
	}
};

ENGINE_TEST(FamilySystem, ChildManagement)

File: Dev/unitTest_Engine_cs/FamilySystem/ChildManagement.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using asd;

namespace unitTest_Engine_cs.FamilySystem
{
	class ChildManagement : EngineTest
	{
		class SpinObject : asd.TextureObject2D
		{
			protected override void OnUpdate()
			{
				Angle += 2;
			}
		}

		class Family
		{
			SpinObject parent, child, childNotAdded;
			Vector2DF position;
			ChildManagementMode mode;

			public SpinObject Parent
			{
				get { return parent; }
			}

			public Family(Vector2DF position, ChildManagementMode mode)
			{
				this.position = position;
				this.mode = mode;
			}

			public void Initialize()
			{
				parent = new SpinObject();
				child = new SpinObject();
				childNotAdded = new SpinObject();

				var texture = Engine.Graphics.CreateTexture2D(CloudTexturePath);
				parent.Texture = texture;
				parent.CenterPosition = texture.Size.To2DF() / 2;
				parent.Scale = new Vector2DF(0.5f, 0.5f);
				parent.Position = position;

				child.Texture = texture;
				child.CenterPosition = texture.Size.To2DF() / 2;
				child.Scale = new Vector2DF(0.5f, 0.5f);
				child.Position = position + new Vector2DF(0, 100);

				childNotAdded.Texture = texture;
				childNotAdded.CenterPosition = texture.Size.To2DF() / 2;
				childNotAdded.Scale = new Vector2DF(0.5f, 0.5f);
				childNotAdded.Position = position + new Vector2DF(0, 200);

				parent.AddChild(child, mode, ChildTransformingMode.Nothing);
				parent.AddChild(childNotAdded, mode, ChildTransformingMode.Nothing);

				Engine.AddObject2D(parent);
				if((mode & ChildManagementMode.RegistrationToLayer) == 0)
				{
					Engine.AddObject2D(child);
				}
			}
		}

		public ChildManagement() : base(60)
		{
		}

		private Family nothing, isDrawnChanged, isUpdateChanged, registrationToLayerChanged, disposed;

		protected override void OnStart()
		{
			nothing = new Family(new Vector2DF(70, 70), ChildManagementMode.Nothing);
			isDrawnChanged = new Family(new Vector2DF(190, 70), ChildManagementMode.IsDrawn);
			isUpdateChanged = new Family(new Vector2DF(310, 70), ChildManagementMode.IsUpdated);
			registrationToLayerChanged = new Family(new Vector2DF(430, 70), ChildManagementMode.RegistrationToLayer);
			disposed = new Family(new Vector2DF(550, 70), ChildManagementMode.Vanishment);

			nothing.Initialize();
			isDrawnChanged.Initialize();
			isUpdateChanged.Initialize();
			registrationToLayerChanged.Initialize();
			disposed.Initialize();
		}

		protected override void OnUpdating()
		{
			if(Time == 30)
			{
				isDrawnChanged.Parent.IsDrawn = false;
				isUpdateChanged.Parent.IsUpdated = false;
				Engine.RemoveObject2D(registrationToLayerChanged.Parent);
				disposed.Parent.Dispose();
			}
		}
	}
}
